cmd/lotus-chainwatch/templates.go

```diff
@@ -0,0 +1,253 @@
+package main
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	rice "github.com/GeertJohan/go.rice"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type handler struct {
+	api    api.FullNode
+	st     *storage
+	site   *rice.Box
+	assets http.Handler
+
+	templates map[string]*template.Template
+}
+
+func newHandler(api api.FullNode, st *storage) (*handler, error) {
+	h := &handler{
+		api:  api,
+		st:   st,
+		site: rice.MustFindBox("site"),
+
+		templates: map[string]*template.Template{},
+	}
+	h.assets = http.FileServer(h.site.HTTPBox())
+
+	funcs := template.FuncMap{
+		"count":    h.count,
+		"countCol": h.countCol,
+		"sum":      h.sum,
+		"netPower": h.netPower,
+		"queryNum": h.queryNum,
+		"sizeStr":  sizeStr,
+		"strings":  h.strings,
+		"messages": h.messages,
+
+		"param": func(string) string { return "" }, // replaced in request handler
+	}
+
+	base := template.New("")
+
+	base.Funcs(funcs)
+
+	return h, h.site.Walk("", func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) != ".html" {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		log.Info(path)
+
+		h.templates["/"+path], err = base.New(path).Parse(h.site.MustString(path))
+		return err
+	})
+}
+
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h, err := newHandler(h.api, h.st) // for faster dev
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	p := r.URL.Path
+	if p == "/" {
+		p = "/index.html"
+	}
+
+	t, ok := h.templates[p]
+	if !ok {
+		h.assets.ServeHTTP(w, r)
+		return
+	}
+
+	t, err = t.Clone()
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	t.Funcs(map[string]interface{}{
+		"param": r.FormValue,
+	})
+
+	if err := t.Execute(w, nil); err != nil {
+		log.Errorf("%+v", err)
+		return
+	}
+
+	log.Info(r.URL.Path)
+}
+
+// Template funcs
+
+func (h *handler) count(table string, filters ...string) (int, error) {
+	// explicitly not caring about sql injection too much, this doesn't take user input
+
+	filts := ""
+	if len(filters) > 0 {
+		filts = " where "
+		for _, filter := range filters {
+			filts += filter + " and "
+		}
+		filts = filts[:len(filts)-5]
+	}
+
+	var c int
+	err := h.st.db.QueryRow("select count(1) from " + table + filts).Scan(&c)
```

Comment:
Can we do somehow without building queries? I know that they should be safe but...

Reply:
That's fine